[docs][windows]: Pod OS field update (#30436)
kubernetes/website: content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md

  * `emptyDir` volumes
  * Named pipe host mounts
  * Resource limits
  * OS field: This `Name` field within OS field should be set to `windows` to indicate that the current pod is a Windows pod. 
              Please note that there is no effect of having value for this field since it is behind a featuegate `IdentifyPodOS` and it is currently alpha.

>               If the `Name` field in `OS` field is set to `windows`, following fields must be unset on the Windows pods when the `IdentifyPodOS` featuregate is enabled:

Reviewer: What's the name of the field that should be set to `windows`? (It's unlikely to be `OS.Name` in the API; the API usually uses camelCase aka lowerCamelCase).

Author: As you noted [here](https://github.com/kubernetes/website/pull/30436#discussion_r755351724), it is spec.os.name

  * `emptyDir` volumes
  * Named pipe host mounts
  * Resource limits
  * OS field: `pod.spec.os.name` should be set to `windows` to indicate that the current pod is a Windows pod. 

>               Please note that there is no effect of having value for this field since it is behind a featuegate `IdentifyPodOS` and it is currently alpha.

Reviewer 2: ```suggestion
              Please note that there is no effect when setting a value for this field since it is behind a feature gate `IdentifyPodOS` and it is currently alpha.
```

if the feature gate is enabled does it have an affect?  It seems from below it does.  Is this field available even when the feature gate is not set? If that is the case, it would be clearer to say something along those lines.

>               Please note that there is no effect when setting a value for this field since it is behind a feature gate `IdentifyPodOS` and it is currently alpha.

Reviewer 2: if the feature gate is enabled does it have an affect? It seems from below it does. Is this field available even when the feature gate is not set? If that is the case, it would be clearer to say something along those lines.

Reviewer 3: can we use the feature-state functionality here for consistency and also to avoid using 'currently' (which may get stale)
{{< feature-state for_k8s_version="v1.23" state="alpha" >}}

Ex: https://github.com/kubernetes/website/blob/269dad43a0d6d7df31bfc9d7b08faeaa866a4f78/content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md?plain=1#L608-L610

Author: Usually all the fields will be available irrespective of the featuregates. We try to drop fields when persisting to storage so that it won't have any effect. 

More info: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-unstable-features-to-stable-versions. Look at the create and update strategy code.

Reviewer 2: What does no effect mean here? This says there is no effect but then goes on to say "the following fields must be unset on the Windows pods when the `IdentifyPodOS` featuregate is enabled" which implies there is a difference in behavior: if you set this field then your pod will be rejected if those fields are set.

Author: > What does no effect mean here?

OS field has no effect when the `IdentifyPodOS` featuregate is disabled. It can be `windows` or `linux` but it doesn't matter however when the `IdentifyPodOS` featuregate is enabled and if linux specific fields are set in the security context of the pod when the name is set to `windows`, pod validation would fail

Reviewer 2: It needs to be clearer that there is no affect (besides the older kubectl version that @sftim mentioned below)  when disabled but when enabled there is an affect (validation could fail). Right now this reads as there is not affect for this field.

Reviewer 2: left a possible suggestion

>               {{< note >}}

Reviewer 3: can we use the feature-state functionality here for consistency and also to avoid using 'currently' (which may get stale)
{{< feature-state for_k8s_version="v1.23" state="alpha" >}}

Ex: https://github.com/kubernetes/website/blob/269dad43a0d6d7df31bfc9d7b08faeaa866a4f78/content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md?plain=1#L608-L610

Author: @marosset - I added it [here](https://github.com/kubernetes/website/pull/30436/files#diff-29fa69c79ca964d4cc4a971d7eadbcfdfe203f97bff1a25057771c7444b147c9R157). PTAL

Reviewer: If you set this field and use an older `kubectl`, I think client-side validation will fail and that does count as an effect. We should make that clear if we can.

Author: How do we mention it usually in our docs?

>               If the `pod.spec.os.name` field is set to `windows`, following fields must be unset on the Windows pods when the `IdentifyPodOS` featuregate is enabled:

Reviewer 2: Are there fields that need to be unset when the Podfield is set to `linux`?  Such as the `securityContext.WindowsOptions`?  Are we documenting that anywhere?

Author: This should be available in the API docs for the field. Since we have a separate documentation for the Windows pods, I am mentioning it explicitly.

Reviewer 2: Thanks, I remember those docs now.  Makes sense to mention it here but I do wonder about maintaining this list in to places.

Author: I am not sure which other place you're recommending.

Reviewer 2: This looks like the same list here: https://github.com/kubernetes/kubernetes/blob/c1153d3353bd4f4b68d85245d53d2745586be474/pkg/apis/core/types.go#L2968-L2990 which will be published as API docs here: https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/ once the does are published.

My main concern here is these two will get out of sync.

Author: So, you don't want this list to be mentioned here in our Windows docs?

Reviewer 2: I think it should be mentioned but there should be one place where all the fields are listed otherwise it will get out sync and be confusing.

  * `emptyDir` volumes
  * Named pipe host mounts
  * Resource limits
  * OS field: 
              {{< feature-state for_k8s_version="v1.23" state="alpha" >}}

>               `pod.spec.os.name` should be set to `windows` to indicate that the current pod is a Windows pod. 

Reviewer: ```suggestion
               `.spec.os.name` should be set to `windows` to indicate that the current Pod uses Windows containers.
```

>               Please note that there is no effect when setting a value for this field since it is behind a feature gate 

Reviewer 2: ```suggestion
              Please note that there is no effect when setting a value for this field when the feature gate 
```

Author: Thank you, I've committed your suggestions.

>               `IdentifyPodOS`.

Reviewer 2: ```suggestion
              `IdentifyPodOS` is not enabled.
```

              {{< note >}}

>               If the `pod.spec.os.name` field is set to `windows`, following fields must be unset on the Windows pods when the `IdentifyPodOS` featuregate is enabled:

Reviewer: ```suggestion
              If the `IdentifyPodOS` feature gate is enabled and you set the `.spec.os.name` field to `windows`, you must not set the following fields in the `.spec` of that Pod:
```
We should also explain what happens if you try.

               * spec.hostPID
               * spec.hostIPC
               * spec.securityContext.seLinuxOptions
               * spec.securityContext.seccompProfile
               * spec.securityContext.fsGroup
               * spec.securityContext.fsGroupChangePolicy
               * spec.securityContext.sysctls
               * spec.shareProcessNamespace
               * spec.securityContext.runAsUser
               * spec.securityContext.runAsGroup
               * spec.securityContext.supplementalGroups
               * spec.containers[*].securityContext.seLinuxOptions
               * spec.containers[*].securityContext.seccompProfile
               * spec.containers[*].securityContext.capabilities
               * spec.containers[*].securityContext.readOnlyRootFilesystem
               * spec.containers[*].securityContext.privileged
               * spec.containers[*].securityContext.allowPrivilegeEscalation
               * spec.containers[*].securityContext.procMount
               * spec.containers[*].securityContext.runAsUser

>                * spec.containers[*].securityContext.runAsGroup

Reviewer: Please put all these field paths in backticks (individually)

  * `emptyDir` volumes
  * Named pipe host mounts
  * Resource limits

>   * OS field: This value should be set to `windows` to indicate that the current pod is a Windows pod.

Reviewer 3: We should either add a note around here saying this functionality is in alpha or link to another page/section that does so.

  node. They should be applied to all containers as a best practice if the operator
  wants to avoid overprovisioning entirely.
* `securityContext.allowPrivilegeEscalation` -

>    not possible on Windows; none of the capabilities are hooked up. Pod API

Reviewer 4: There are other places in the doc that talks about these fields under `securityContext`. Shouldn't we also mention it there? Or is that not done for features behind a feature gate?

>    validation would fail when `IdentifyPodOS` featuregate is enabled.

Reviewer 3: I think for now it might be less confusing if we left this section as is and added a new section just for describing the new behavior if `IdenfityPodOS`. 
Also these checks only happen if both the feature gate is enabled and the pod specifies `OS=windows` which i think should be called out as well. The current updates could cause some confusion if only the feature gate was enabled but the OS isn't specified.

* `securityContext.capabilities` -
   POSIX capabilities are not implemented on Windows. Pod API
   validation would fail when `IdentifyPodOS` featuregate is enabled.
* `securityContext.privileged` -
   Windows doesn't support privileged containers. Pod API
   validation would fail when `IdentifyPodOS` featuregate is enabled.
* `securityContext.procMount` -
   Windows doesn't have a `/proc` filesystem. Pod API
   validation would fail when `IdentifyPodOS` featuregate is enabled.
* `securityContext.readOnlyRootFilesystem` -
   not possible on Windows; write access is required for registry & system
   processes to run inside the container. Pod API
   validation would fail when `IdentifyPodOS` featuregate is enabled.
* `securityContext.runAsGroup` -
   not possible on Windows as there is no GID support. Pod API
   validation would fail when `IdentifyPodOS` featuregate is enabled.
* `securityContext.runAsNonRoot` -
   Windows does not have a root user. The closest equivalent is `ContainerAdministrator`

>    which is an identity that doesn't exist on the node. 

Reviewer 2: Surprised this isn't updated against this base: https://github.com/kubernetes/website/pull/30256/files

* `securityContext.runAsUser` -
   use [`runAsUsername`](/docs/tasks/configure-pod-container/configure-runasusername)
   instead. Pod API

>    validation would fail when `IdentifyPodOS` featuregate is enabled.

Reviewer 2: can we re-order these two sentences. When I read it I thought we said `runAsUsername` would fail

The following list documents differences between how Pod specifications work between Windows and Linux:

* `hostIPC` and `hostpid` - host namespace sharing is not possible on Windows. Pod API
   validation would fail when `IdentifyPodOS` featuregate is enabled.   
* `hostNetwork` - There is no Windows OS support to share the host network
* `dnsPolicy` - setting the Pod `dnsPolicy` to `ClusterFirstWithHostNet` is
   not supported on Windows because host networking is not provided. Pods always
   run with a container network.
* `podSecurityContext` (see below)
* `shareProcessNamespace` - this is a beta feature, and depends on Linux namespaces
  which are not implemented on Windows. Windows cannot share process namespaces or

>   the container's root filesystem. Only the network can be shared. Pod API

Reviewer 2: side note, with HPC we do share the root filesystem not sure if that is worth noting since it is out of context of shared namespaces
